src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,466 @@
+                        assert 'string' in field['type']['items']
+                        default_value = [None]
+                object_[field['name']] = default_value
+
+    @classmethod
+    def replace_null_with_empty_string(cls, object_json: MutableJSON):
+        # FIXME: remove with https://github.com/DataBiosphere/azul/issues/2462
+        for k, v in object_json.items():
+            if isinstance(v, dict):
+                cls.replace_null_with_empty_string(v)
+            elif v is None:
+                object_json[k] = ''
+            elif v == [None]:
+                object_json[k] = ['']
+
+    def to_json(self, relations: Iterable[PFBRelation]):
+        return {
+            **attr.asdict(self),
+            "relations": [attr.asdict(relation) for relation in relations]
+        }
+
+
+class PFBTransformer:
+    """
+    Transform documents from ElasticSearch into PFB entities. A document's inner
+    entities correspond to PFB entities which are normalized and linked via
+    Relations.
+    """
+
+    def __init__(self, schema: JSON):
+        self.schema = schema
+        self._entities_to_relations: MutableMapping[PFBEntity,
+                                                    MutableSet[PFBRelation]] = defaultdict(set)
+
+    def add_doc(self, doc: JSON):
+        """
+        Add an Elasticsearch document to be transformed.
+        """
+        contents = doc['contents']
+        # File entities are assumed to be unique
+        file_entity = PFBEntity.from_json(name='files',
+                                          object_=one(contents['files']),
+                                          schema=self.schema)
+        assert file_entity not in self._entities_to_relations
+        # File entities won't relate *to* anything, but things relate *to* them.
+        # Terra streams PFBs and thus entities must be defined before they are
+        # referenced. This is one reason that file_entity is added first.
+        self._entities_to_relations[file_entity] = set()
+        for entity_type, entity in contents.items():
+            if entity_type != 'files':
+                if len(entity) == 0:
+                    continue
+                elif len(entity) == 1:
+                    entity = PFBEntity.from_json(name=entity_type,
+                                                 object_=one(entity),
+                                                 schema=self.schema)
+                    self._entities_to_relations[entity].add(PFBRelation.to_entity(file_entity))
+                else:
+                    assert False, (doc, entity_type)
+
+    def entities(self) -> Iterable[PFBEntity]:
+        for entity, relations in self._entities_to_relations.items():
+            # Sort relations to make entities consistent for easy diffing
+            yield entity.to_json(sorted(relations, key=attrgetter('dst_name', 'dst_id')))
+
+
+def write_pfb_entities(entities: Iterable[JSON], json_schema: JSON, path: str):
```

Comment:
This seems to be the main entry point, so it should come first in the module. And be sure to use the appropriate naming conventions for internal functions and protected methods. If it's not clear why this is desirable, I can explain on PL.